core/shared/platform/zephyr/zephyr_socket.c

```diff
@@ -0,0 +1,1049 @@
+        addr->sin_port = htons(bh_sockaddr->port);
+        addr->sin_family = AF_INET;
+        addr->sin_addr.s_addr = htonl(bh_sockaddr->addr_buffer.ipv4);
+        *socklen = sizeof(*addr);
+    }
+#ifdef IPPROTO_IPV6
+    else {
+        struct sockaddr_in6 *addr = (struct sockaddr_in6 *)sockaddr;
+        size_t i;
+        addr->sin6_port = htons(bh_sockaddr->port);
+        addr->sin6_family = AF_INET6;
+
+        for (i = 0; i < sizeof(bh_sockaddr->addr_buffer.ipv6)
+                            / sizeof(bh_sockaddr->addr_buffer.ipv6[0]);
+             i++) {
+            uint16 part_addr = htons(bh_sockaddr->addr_buffer.ipv6[i]);
+            addr->sin6_addr.s6_addr[i * 2] = 0xff & part_addr;
+            addr->sin6_addr.s6_addr[i * 2 + 1] = (0xff00 & part_addr) >> 8;
+        }
+
+        *socklen = sizeof(*addr);
+    }
+#endif
+}
+
+static int
+getaddrinfo_error_to_errno(int error)
+{
+    switch (error) {
+        case DNS_EAI_AGAIN:
+            return EAGAIN;
+        case DNS_EAI_FAIL:
+            return EFAULT;
+        case DNS_EAI_MEMORY:
+            return ENOMEM;
+        case DNS_EAI_SYSTEM:
+            return errno;
+        default:
+            return EINVAL;
+    }
+}
+
+static int
+is_addrinfo_supported(struct zsock_addrinfo *info)
+{
+    return
+        // Allow only IPv4 and IPv6
+        (info->ai_family == AF_INET || info->ai_family == AF_INET6)
+        // Allow only UDP and TCP
+        && (info->ai_socktype == SOCK_DGRAM || info->ai_socktype == SOCK_STREAM)
+        && (info->ai_protocol == IPPROTO_TCP
+            || info->ai_protocol == IPPROTO_UDP);
+}
+
+static int
+os_socket_setbooloption(bh_socket_t socket, int level, int optname,
+                        bool is_enabled)
+{
+    int option = (int)is_enabled;
+
+    if (zsock_setsockopt(socket->fd, level, optname, &option, sizeof(option))
+        != 0) {
+        return BHT_ERROR;
+    }
+
+    return BHT_OK;
+}
+
+static int
+os_socket_getbooloption(bh_socket_t socket, int level, int optname,
+                        bool *is_enabled)
+{
+    assert(is_enabled);
+
+    int optval;
+    socklen_t optval_size = sizeof(optval);
+
+    if (zsock_setsockopt(socket->fd, level, optname, &optval, optval_size)
+        != 0) {
+        return BHT_ERROR;
+    }
+    *is_enabled = (bool)optval;
+    return BHT_OK;
+}
+
+// Platform API implementation
+int
+os_socket_create(bh_socket_t *sock, bool is_ipv4, bool is_tcp)
+{
+    int af = is_ipv4 ? AF_INET : AF_INET6;
+    // now socket is a struct try  *(sock)->fd
+
+    *(sock) = BH_MALLOC(sizeof(zephyr_handle));
+
+    if (!sock) {
+        return BHT_ERROR;
+    }
+
+    if (is_tcp) {
+        (*sock)->fd = zsock_socket(af, SOCK_STREAM, IPPROTO_TCP);
+    }
+    else {
+        (*sock)->fd =
+            zsock_socket(af, SOCK_DGRAM, IPPROTO_UDP); // IPPROTO_UDP or 0 ?
+    }
+
+    (*sock)->is_sock = true;
+
+    return ((*sock)->fd == -1) ? BHT_ERROR : BHT_OK;
```

Comment:
Should BH_FREE(*sock) if return BH_ERROR since the caller may not free it?
```C
    if ((*sock)->fd == -1) {
        BH_FREE(*sock);
        return BHT_ERROR;
    }
    return BHT_OK;
}
```